src/components/kubectl/logs.ts

```diff
@@ -0,0 +1,184 @@
+import { isPod } from '../../kuberesources.objectmodel';
+import * as kuberesources from '../../kuberesources';
+import { ResourceNode } from '../../explorer';
+import * as yaml from 'js-yaml';
+import * as kubectlUtils from '../../kubectlUtils';
+import { host } from '../../host';
+
+export enum LogsDisplayMode {
+    Show,
+    Follow
+}
+
+const FOLLOW_LOGS_TERMINAL = "Kubernetes: Follow Logs";
+
+/**
+ * Fetches logs for a Pod. Handles use cases for fetching pods
+ * from an open document, or from the current namespace.
+ * @param kubectl
+ * @param explorerNode
+ * @param displayMode
+ */
+export async function logsKubernetes(
+    kubectl: Kubectl,
+    explorerNode: ResourceNode | undefined,
+    displayMode: LogsDisplayMode
+) {
+    if (explorerNode) {
+        return await getLogsForExplorerPod(kubectl, explorerNode, displayMode);
+    }
+
+    return findPod(kubectl, displayMode);
+}
+
+/**
+ * Fetch logs from a Pod, when selected from the Explorer.
+ * @param kubectl
+ * @param explorerNode
+ * @param displayMode
+ */
+async function getLogsForExplorerPod(
+    kubectl: Kubectl,
+    explorerNode: ResourceNode,
+    displayMode: LogsDisplayMode
+) {
+    const namespace = explorerNode.namespace;
+    const podSummary = { name: explorerNode.id, namespace: namespace };
+    const container = await selectContainerForPod(podSummary);
+
+    // No container. How'd we get here. How does this Pod even work. What is going on.
+    if (!container) {
+        return;
+    }
+
+    getLogsForContainer(
+        kubectl,
+        podSummary,
+        container.name,
+        displayMode
+    );
+
+    return;
+}
+
+/**
+ * Fetches logs for a pod. If there are more than one containers,
+ * prompts the user for which container to fetch logs for.
+ * @param kubectl
+ * @param podSummary
+ * @param displayMode
+ */
+async function getLogsForPod(kubectl: Kubectl, podSummary: PodSummary, displayMode: LogsDisplayMode) {
+    if (!podSummary) {
+        vscode.window.showErrorMessage('Can\'t find a pod!');
+        return;
+    }
+
+    const container = await selectContainerForPod(podSummary);
+    if (!container) {
+        return;
+    }
+
+    getLogsForContainer(kubectl, podSummary, container.name, displayMode);
+}
+
+/**
+ * Gets the logs for a container in a provided pod, in a provided namespace, in a provided container.
+ * @param kubectl
+ * @param podSummary
+ * @param containerName
+ * @param displayMode
+ */
+function getLogsForContainer(
+    kubectl: Kubectl,
+    podSummary: PodSummary,
+    containerName: string | undefined,
+    displayMode: LogsDisplayMode
+) {
+    let cmd = `logs ${podSummary.name}`;
+
+    if (podSummary.namespace) {
+        cmd = `${cmd} --namespace=${podSummary.namespace}`;
+    }
+
+    if (containerName) {
+        cmd = `${cmd} --container=${containerName}`;
+    }
+
+    if (displayMode === LogsDisplayMode.Follow) {
+        cmd = `${cmd} -f`;
+        kubectl.invokeInNewTerminal(cmd, FOLLOW_LOGS_TERMINAL);
```

Comment:
Suggest follow terminals be labelled as to which pod/container they're following - otherwise you could have several of them and not know which is which.

Reply:
👍